### PR TITLE
[YAML] Schemify avroio and add a yaml provider.

### PR DESCRIPTION
**Pull request by robertwb**
This fixes https://github.com/apache/beam/issues/28852

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://github.com/apache/beam/blob/master/CONTRIBUTING.md#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

### CODE REVIEW

**Comment by robertwb** [comment]
R: @riteshghorse

**Comment by robertwb** [comment]
Run Portable_Python PreCommit

File: sdks/python/apache_beam/io/avroio.py [hunk 1]
```diff
@@ -338,12 +351,15 @@ def split_points_unclaimed(stop_position):
           yield record
 
 
+_create_avro_source = _FastAvroSource
```
Comment (riteshghorse) [inline sdks/python/apache_beam/io/avroio.py:354]: I guess this is for not breaking imports in other files. Looks like it is in these two files: bigquery.py and bigquery_read_internal.py. So we may able to get rid of it. Not blocking though!

**Review by riteshghorse — APPROVED** [review]
LGTM, thanks! Left some comments but nothing blocking.

**Review by Polber — APPROVED** [review]

File: sdks/python/apache_beam/io/avroio.py (hunk 1 above)
Comment (robertwb) [reply sdks/python/apache_beam/io/avroio.py:354]: Yep, exactly.

File: sdks/python/apache_beam/io/avroio.py [hunk 2]
```diff
@@ -382,17 +398,32 @@ def __init__(
...
+      avro_schema = self._schema
+      records = pcoll
+    else:
+      try:
+        beam_schema = schemas.schema_from_element_type(pcoll.element_type)
+      except TypeError as exn:
+        raise ValueError(
+            "An explicit schema is required to write non-schema'd PCollections."
+        ) from exn
+      avro_schema = beam_schema_to_avro_schema(beam_schema)
```
[the remark this answers is not in the document]
Comment (robertwb) [reply sdks/python/apache_beam/io/avroio.py:422]: We also use the absence of an explicit schema to do the pre-processing. Typically expand is only called once, so caching is not a big savings here anyway.

[the remark this answers is not in the document]
Comment (riteshghorse) [reply sdks/python/apache_beam/io/avroio.py:422]: got it, thanks!

**Pull request merged**